dtm0/drlink.c

```diff
@@ -0,0 +1,532 @@
+{
+	const struct m0_rpc_item *item = &drf->df_rfop->f_item;
+	M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, m0_sm_id_get(&drf->df_gen.fo_sm_phase),
+		     m0_sm_id_get(&item->ri_sm));
+}
+
+/*
+ * Establish a relation between a FOM (for example, a CAS PUT FOM).
+ * and a DTM RPC link FOM that was created to send a DTM message:
+ *   FOM <-> DRLINK FOM.
+ */
+static void drlink_addb_drf2parent_relate(struct drlink_fom *drf)
+{
+	M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, drf->df_parent_sm_id,
+		     m0_sm_id_get(&drf->df_gen.fo_sm_phase));
+}
+
+#define F M0_CO_FRAME_DATA
+static void drlink_coro_fom_tick(struct m0_co_context *context)
+{
+	int                 rc   = 0;
+	struct drlink_fom  *drf  = M0_AMB(drf, context, df_co);
+	struct m0_fom      *fom  = &drf->df_gen;
+
+	M0_CO_REENTER(context,
+		      struct m0_long_lock_link   llink;
+		      struct m0_long_lock_addb2  llock_addb2;
+		      struct dtm0_process       *proc;
+		      );
+
+	drlink_addb_drf2parent_relate(drf);
+
+	m0_mutex_lock(&drf->df_svc->dos_generic.rs_mutex);
+	rc = find_or_add(drf->df_svc, &drf->df_tgt, &F(proc));
+	/*
+	 * Safety: it is safe to release the lock right here because
+	 * ::dtm0_service_conns_term is supposed to be called only
+	 * after all drlink FOMs have reached DRF_DONE state.
+	 * There are no other places where processes get removed from the list.
+	 */
+	m0_mutex_unlock(&drf->df_svc->dos_generic.rs_mutex);
+
+	if (rc != 0)
+		goto out;
+
+	m0_long_lock_link_init(&F(llink), fom, &F(llock_addb2));
+
+	M0_CO_FUN(context, co_long_write_lock(context,
+					      &F(proc)->dop_llock,
+					      &F(llink),
+					      DRF_LOCKING));
+	M0_ASSERT(m0_long_is_write_locked(&F(proc)->dop_llock, fom));
+
+	if (dpr_state_infer(F(proc)) == DPR_TRANSIENT) {
+		rc = dtm0_process_rlink_reinit(F(proc), drf);
+		if (rc != 0)
+			goto unlock;
+		M0_CO_FUN(context, co_rpc_link_connect(context,
+						       &F(proc)->dop_rlink,
+						       fom, DRF_CONNECTING));
+	}
+
+	if (dpr_state_infer(F(proc)) == DPR_FAILED)
+			goto unlock;
+
+	M0_ASSERT(dpr_state_infer(F(proc)) == DPR_ONLINE);
+	m0_fom_phase_set(fom, DRF_SENDING);
+	rc = dtm0_process_rlink_send(F(proc), drf);
+	if (rc != 0)
+		goto unlock;
+
+	/* Safety: FOP (and item) can be released only in ::drlink_fom_fini. */
+	drlink_addb_drf2item_relate(drf);
+
+	if (drf->df_wait_for_ack) {
+		M0_CO_YIELD_RC(context,
+			       m0_co_op_tick_ret(&drf->df_co_op, fom,
+						 DRF_WAITING_FOR_REPLY));
+		m0_co_op_reset(&drf->df_co_op);
+		rc = m0_rpc_item_error(&drf->df_rfop->f_item);
+	}
+
+unlock:
+	m0_long_write_unlock(&F(proc)->dop_llock, &F(llink));
+	m0_long_lock_link_fini(&F(llink));
+out:
+	if (rc != 0)
+		m0_fom_phase_move(fom, rc, DRF_FAILED);
+
+	m0_fom_phase_set(fom, DRF_DONE);
+}
+#undef F
+
+static int drlink_fom_tick(struct m0_fom *fom)
+{
+	struct m0_co_context *co = &fom2drlink_fom(fom)->df_co;
+	M0_CO_START(co);
+	drlink_coro_fom_tick(co);
+	return M0_CO_END(co) ?: M0_FSO_WAIT;
```

Comment:
A fo_tick() function can only return M0_FSO_WAIT or M0_FSO_AGAIN.
No other value is accepted.
Please note, here error code can be returned from M0_CO_END(), which will cause panic.

Reply:
M0_CO_END may return 0, M0_FSO_WAIT, or M0_FSO_AGAIN. This expression is needed to convert 0 into M0_FSO_WAIT, so that the result of this function will be either M0_FSO_WAIT or M0_FSO_AGAIN.

I updated the code and added a comment to make it more obvious in https://github.com/Seagate/cortx-motr/pull/802/commits/55385c7ee5f01d4d021d647f61f61f7a497ae224